security/pfSense-pkg-nmap/files/usr/local/www/nmap_scan.php

```diff
@@ -0,0 +1,468 @@
+	return $valid_str;
+}
+
+$input_errors = array();
+$do_nmapscan = false;
+if ($_POST) {
+	$hostname = $_POST['hostname'];
+	$interface = $_POST['interface'];
+	$scantype = $_POST['scantype'];
+	$noping = isset($_POST['noping']);
+	$servicever = isset($_POST['servicever']);
+	$osdetect = isset($_POST['osdetect']);
+	$custom_scantype = $_POST['custom_scantype'];
+	$custom_options = $_POST['custom_options'];
+	$custom_targetspecs = $_POST['custom_targetspecs'];
+
+	if ($_POST['startbtn'] != "") {
+		$action = gettext("Start");
+
+		// check for input errors
+		if (empty($hostname)) {
+			$input_errors[] = gettext("You must enter an IP address to scan.");
+		} elseif (!(is_ipaddr($hostname) || is_subnet($hostname) || is_hostname($hostname))) {
+			$input_errors[] = gettext("You must enter a valid IP address to scan.");
+		}
+
+		if(!empty($interface)) {
+			if (!array_key_exists($interface, $interfaces)) {
+				$input_errors[] = gettext("Invalid interface.");
+			}
+		}
+
+		// process scan options
+		if (!count($input_errors)) {
+
+			// prevent use of any deprecated option
+			$nmap_options = " -d";
+
+			if (is_ipaddrv6($hostname) || is_subnetv6($hostname)) {
+				$nmap_options .= " -6";
+			}
+
+			// scan type
+			if (empty($custom_scantype)) {
+				switch($scantype) {
+					case 'syn':
+						$nmap_options .= " -sS";
+						break;
+					case 'connect':
+						$nmap_options .= " -sT";
+						break;
+					case 'icmp':
+						$nmap_options .= " -sn"; // previously -sP
+						break;
+					case 'udp':
+						$nmap_options .= " -sU";
+						break;
+					case 'arp':
+						$nmap_options .= " -sn -PR";
+						break;
+				}
+			} elseif (valid_input($custom_scantype)) {
+					$nmap_options .= " " . $custom_scantype;
+			} else {
+					$input_errors[] = gettext("Invalid characters in Custom Scan Type. Only alphanumeric, spaces and '.,-_' chars are allowed. Custom output options -o are not allowed");
+			}
+
+			// scan options
+			if ($noping) {
+				$nmap_options .= " -P0";
+			}
+			if ($servicever) {
+				$nmap_options .= " -sV";
+			}
+			if ($osdetect) {
+				$nmap_options .= " -O";
+			}
+
+			if (!empty($custom_options)) {
+				if (valid_input($custom_options)) {
+					$nmap_options .= " " . $custom_options;
+				} else {
+					$input_errors[] = gettext("Invalid characters in custom options. Only alphanumeric, spaces and '.,-_' chars are allowed. Custom output options -o are not allowed");
+				}
+			}
+
+			// append summary output to results file (doesn't contain stderr)
+			// prevent using source stylesheets for xls output (not really needed)
+			$nmap_options .= " -oN {$fp}{$fn} --append-output --no-stylesheet";
+
+			if (!empty($interface)) {
+				$nmap_options .= " -e " . get_real_interface($interface);
+			}
+
+			if (!empty($custom_targetspecs)) {
+				if (valid_input($custom_targetspecs)) {
+					$nmap_options .= " " . $custom_targetspecs;
```

Comment:
Same note as above about arbitrary parameters.

Reply:
> I know you're severely limiting the input allowed here but it still seems like a bad idea to allow arbitrary parameters to be passed.
> 
> If someone knows that much about nmap parameters they aren't likely to be using the GUI for this. Possible, but unlikely.

> Same note as above about arbitrary parameters.


Thank you a lot for the detailed input.
Like mentioned in the redmine comment, I am aware that there is a potential risk if nmap ever changes the options during an update, or in case of nmap vulnerabilities. Adding changes for managing user permissions on the custom options seems overkill in fact for just nmap. As you say, any admin user could just drop to ssh/shell in that case.

If you agree, I'd go with simply removing the unsafe custom options fields. I could add a few additional selectable common options and maybe one or two more scan options. The hosts field could also be adjusted to allow comma or space separated IP/hosts.

Also, I'd like your input about this part:
https://github.com/pfsense/FreeBSD-ports/blob/654af55a3208ea8c9aacf8f0a2618ab52f74206f/security/pfSense-pkg-nmap/files/usr/local/www/nmap_scan.php#L86
```
	if ($_POST['startbtn'] != "") {
		$action = gettext("Start");

		// check for input errors
		if (empty($hostname)) {
```

Currently, in "Packet Capture" module, any button like "Clear" or "View Results" will check for input errors and do not allow processing.
I went with the option to only check for fields integrity when doing a scan. The fields are still processed and thus preserved after clicking View, Clear or Refresh buttons. There are no additional risks, especially if custom options are removed. Since they are never used except for preserving the form print, I found it odd that click "clear" buttons fails if invalid entries are entered.

Also, currently Packet Capture stops if page is refreshed. If the changes made here are fine, I can quickly modify Packet Capture GUI to:
- properly run in background even if user refreshes the page
- implement the Refresh results button while a scan is running
- remove check forms integrity when user clicks other buttons except Scan

Reply:
The packet capture page doesn't stop on refresh, it runs until the capture stops -- when it hits an upper limit on packets. If you start a long capture, say for lots of packets with a filter that doesn't match a lot, then it will keep running for a while. You'll still get a stop button even if you navigate away and come back.

Reply:
> The packet capture page doesn't stop on refresh, it runs until the capture stops -- when it hits an upper limit on packets. If you start a long capture, say for lots of packets with a filter that doesn't match a lot, then it will keep running for a while. You'll still get a stop button even if you navigate away and come back.

Ah, yes, you're right, it was because I tested on WAN and it reached the 100 hits in a few seconds

I can remove the custom fields and submit a new commit ?

Reply:
It is safer to remove the custom fields, if ever nmap changes again the -o option for example as they previously did
The below commits look fine to me and I tested all possibilities I could figure